### PR TITLE
[NETBEANS-1763] Integrate maven failsafe output into test runner output

### PR DESCRIPTION
**Pull request by matthiasblaesing**
This PR reworks the junit integration, so that both the surefire (unit tests) and
failsafe (integration tests) plugin are supported. The output of both plugins
is mostly identical, so does not need special handling for displaying test results,
to support rerunning integration tests, new actions are needed and implemented.

### CODE REVIEW

**Comment by matthiasblaesing** [comment]
@jmborer you raised https://issues.apache.org/jira/browse/NETBEANS-1763 it would be great if you could have a look at this PR and test it.

**Comment by matthiasblaesing** [comment]
If no-one objects, I'll merge this sometime next week.

File: java/maven/src/org/netbeans/modules/maven/ActionProviderImpl.java [hunk 1]
```diff
@@ -113,11 +113,13 @@
...
-    public static final String RUN_MAIN = ActionProvider.COMMAND_RUN_SINGLE + ".main";
-    public static final String DEBUG_MAIN = ActionProvider.COMMAND_DEBUG_SINGLE + ".main";
-    public static final String PROFILE_MAIN = ActionProvider.COMMAND_PROFILE_SINGLE + ".main";
+
+    public static final String COMMAND_RUN_MAIN = ActionProvider.COMMAND_RUN_SINGLE + ".main";
+    public static final String COMMAND_DEBUG_MAIN = ActionProvider.COMMAND_DEBUG_SINGLE + ".main";
+    public static final String COMMAND_PROFILE_MAIN = ActionProvider.COMMAND_PROFILE_SINGLE + ".main";
+
+    public static final String COMMAND_INTEGRATION_TEST_SINGLE = "integration-test.single";
+    public static final String COMMAND_DEBUG_INTEGRATION_TEST_SINGLE = "debug.integration-test.single";
```
Comment (junichi11) [inline java/maven/src/org/netbeans/modules/maven/ActionProviderImpl.java]: Add `// NOI18N` ?

File: java/maven.junit/src/org/netbeans/modules/maven/junit/JUnitOutputProcessorFactory.java [hunk 2]
```diff
@@ -45,7 +45,10 @@ public JUnitOutputProcessorFactory() {
 
     @Override
     public Set<OutputProcessor> createProcessorsSet(Project project, RunConfig config) {
-        if (config.getGoals().contains("test") || config.getGoals().contains("integration-test") || config.getGoals().contains("surefire:test")) { //NOI18N
+        if (config.getGoals().contains("test") //NOI18N
+            || config.getGoals().contains("integration-test") //NOI18N
+            || config.getGoals().contains("surefire:test")
```
Comment (junichi11) [inline java/maven.junit/src/org/netbeans/modules/maven/junit/JUnitOutputProcessorFactory.java]: Missing `// NOI18N`

File: java/maven.junit/src/org/netbeans/modules/maven/junit/JUnitOutputListenerProvider.java [hunk 3]
```diff
@@ -244,42 +247,69 @@ static boolean isFullJavaId(String possibleNewRunningTestClass) {
...
-		}
-		if (null != outputDir) {
-		    createSession(outputDir);
-		}
-	    }
-	}
+        reportNameSuffix = null;
+        testType = null;
+        String reportsDirectory = null;
+	if (("mojo-execute#surefire:test".equals(sequenceId))) {
```
Comment (junichi11) [inline java/maven.junit/src/org/netbeans/modules/maven/junit/JUnitOutputListenerProvider.java]: Missing `NOI18N`s?

File: java/maven.junit.ui/src/org/netbeans/modules/maven/junit/ui/MavenJUnitTestMethodNode.java [hunk 4]
```diff
@@ -62,52 +63,71 @@ public MavenJUnitTestMethodNode(Testcase testcase, Project project, String proje
...
                                 nodeContext,
                                 COMMAND_DEBUG_SINGLE_METHOD,
                                 Bundle.LBL_DebugTest()));
+                        }
+                        if (integrationTest
+                            && action.equals("integration-test.single")
+                            && actionProvider.isActionEnabled(action, nodeContext)) {
+                            actions.add(new TestMethodNodeAction(actionProvider,
+                                nodeContext,
+                                "integration-test.single",
```
Comment (junichi11) [inline java/maven.junit.ui/src/org/netbeans/modules/maven/junit/ui/MavenJUnitTestMethodNode.java]: Missing `// NOI18N`s?

File: java/maven.junit.ui/src/org/netbeans/modules/maven/junit/ui/MavenJUnitTestMethodNode.java [hunk 5]
```diff
@@ -62,52 +63,71 @@ public MavenJUnitTestMethodNode(Testcase testcase, Project project, String proje
...
         List<Action> actions = new ArrayList<Action>();
-        Action preferred = getPreferredAction();
-        if (preferred != null) {
-            actions.add(preferred);
-        }
-        FileObject testFO = getTestcaseFileObject();                            
-        if (testFO != null){
+        FileObject testFO = getTestcaseFileObject();
+        if (testFO != null) {
+            boolean unitTest = getTestcase().getType() == null || "UNIT".equals(getTestcase().getType());
```
Comment (junichi11) [inline java/maven.junit.ui/src/org/netbeans/modules/maven/junit/ui/MavenJUnitTestMethodNode.java]: Missing `// NOI18N`s?

**Comment by matthiasblaesing** [comment]
@junichi11 thanks for looking into this - I updated the missing `NOI18N` marker and updated the commits in place. Clear to merge?

File: java/maven.junit.ui/src/org/netbeans/modules/maven/junit/ui/MavenJUnitTestMethodNode.java [hunk 6]
```diff
@@ -62,52 +63,71 @@ public MavenJUnitTestMethodNode(Testcase testcase, Project project, String proje
...
+                        if (unitTest
+                            && action.equals(COMMAND_DEBUG_SINGLE_METHOD)
+                            && actionProvider.isActionEnabled(action, nodeContext)) {
+                            actions.add(new TestMethodNodeAction(actionProvider,
                                 nodeContext,
                                 COMMAND_DEBUG_SINGLE_METHOD,
                                 Bundle.LBL_DebugTest()));
+                        }
+                        if (integrationTest
+                            && action.equals("integration-test.single")
```
Comment (junichi11) [inline java/maven.junit.ui/src/org/netbeans/modules/maven/junit/ui/MavenJUnitTestMethodNode.java]: Missing 4 `NOI18N`s

Maybe, should add `NOI18N` to `aName = "";` and `bName = "";`

**Comment by junichi11** [comment]
@matthiasblaesing I'm not an expert of this area, but if there are steps to verify this, I would like to do it :)
@jmborer Could you please take a look at this if possible (because you are the reporter as Matthias wrote)?

**Comment by jmborer** [comment]
Hi, I am currently on vacation and can’t therefore not properly have a look at this. 

Question (stupid one) how can I test the code myself? I am quite new to large oss projects on git. 

Thank you for your help.

**Comment by junichi11** [comment]
@jmborer Thank you for your reply.
> Question (stupid one) how can I test the code myself?

#### How to build
See https://github.com/apache/netbeans/blob/master/README.md

#### How to apply this patch
Maybe, there are some ways, I suppose. 
One of the simplest way:
```sh
git clone git@github.com:apache/netbeans.git
# or git clone https://github.com/apache/netbeans.git
cd netbeans
git checkout -b pr1402
curl -L https://github.com/apache/netbeans/pull/1402.patch | git am
```
BTW, you are on vacation. So, when are you going to do it?

Thanks.

**Comment by jmborer** [comment]
I’ll be back on 20th of August. I’ll try to test on 21 or 22. Is this OK?

Envoyé de mon iPad

**Comment by matthiasblaesing** [comment]
@jmborer if you are looking for a binary, you can get this one: https://doppel-helix.eu/netbeans-dev.zip That is the build I'm currently using at work. it is based on commit 39e97b900a8d604f3968e07605ee0d7709f37977 with PR #1417. The timeframe is totally ok.

@junichi11 the PRs intentend change is visible, when you run a maven project, that has an integration test defined. This is a sample project:
[testTests.zip](https://github.com/apache/netbeans/files/3506166/testTests.zip)
Open it, run the custom goal "integration-test" and open the test-results window. You will see one output with the unittest (on current netbeans), however, when you look at the console, you see that there are in fact two tests run (one unittest, one integration test).
With this PR, a second test results window is opened, that shows the integration test.

**Comment by junichi11** [comment]
@matthiasblaesing Thank you! I'll give it a try later.

**Comment by junichi11** [comment]
Verified. It seems to work fine :)
netbeans-pr-1402

**Comment by toepi** [comment]
sry to ask, has this change also a impact on 'Test File'? Currently if I you it on a `*Test` file it will use surefire to run the test cases and all is fine, but if I use it in a `*IT` file it will also use surefire, and so the configuration for failsafe will not be used and bad thinks happens ... next nice to have will be execution integrationtests should also run plugins in pre- and post-integrationtest-phase e.g. start stop some external stuff?

**Comment by matthiasblaesing** [comment]
@toepi no it does not change the GUI integration for running test. That was out of scope for this. For running post/pre integration things: This hooks into the maven build machine, so if maven would do it, netbeans will do it.

**Comment by matthiasblaesing** [comment]
@junichi11 thanks for checking

**Comment by Saljack** [comment]
Would it be possible to add something like IT into tab label instead of `_2` for example `nameofproject... IT`?

**Comment by matthiasblaesing** [comment]
@Saljack I updated the pr and it now contains information about the test type:

Bildschirmfoto_2019-08-25_13-23-18

**Comment by jmborer** [comment]
Hi guys, 
I had a extremely busy return of vacation. I am currently trying to download matthiasblaesing compiled version and will give it a go to check the issue.

**Comment by jmborer** [comment]
@matthiasblaesing:  Unfortunately, I am not able to get your binary. I am stuck behind a very annoying corporate proxy that prevents a lot of stuff.

It will take me a lot of time to get the dev version compiled in our bizz environment with your fix. So, I am confident that the fix is working. So far you have my go for the PR.

**Comment by matthiasblaesing** [comment]
Thank you all for having a look - will merge. If problems are found later, I'm still around :-)

**Pull request merged**